src/dhcpmon/src/dhcp_device.c

```diff
@@ -0,0 +1,435 @@
+#define DHCP_BPF_COMPILE_OPTIMIZE 1
+/** DHCP BPF libpcap netmask */
+#define DHCP_BPF_NETMASK 0
+/** Start of Ether header of a captured frame */
+#define ETHER_START_OFFSET  0
+/** Start of IP header of a captured frame */
+#define IP_START_OFFSET (ETHER_START_OFFSET + ETHER_HDR_LEN)
+/** Start of UDP header of a captured frame */
+#define UDP_START_OFFSET (IP_START_OFFSET + sizeof(struct ip))
+/** Start of DHCP header of a captured frame */
+#define DHCP_START_OFFSET (UDP_START_OFFSET + sizeof(struct udphdr))
+/** Start of DHCP Options segment of a captured frame */
+#define DHCP_OPTIONS_HEADER_SIZE 240
+
+/**  DHCP Berkley Packet Filter */
+static const char *DHCP_BPF = "udp and (port 67 or port 68)";
+
+/** global aggregate counter for DHCP interfaces */
+static dhcp_device_counters_t glob_counters[DHCP_DIR_COUNT] = {
+    [DHCP_RX] = { .discovery = 0, .offer = 0, .request = 0, .ack = 0 },
+    [DHCP_TX] = { .discovery = 0, .offer = 0, .request = 0, .ack = 0 },
+};
+
+/** global aggregate counter snapshot for DHCP interfaces */
+static dhcp_device_counters_t glob_counters_snapshot[DHCP_DIR_COUNT] = {
+    [DHCP_RX] = { .discovery = 0, .offer = 0, .request = 0, .ack = 0 },
+    [DHCP_TX] = { .discovery = 0, .offer = 0, .request = 0, .ack = 0 },
+};
+
+
+/**
+ * @code handle_dhcp_option_53(context, dhcp_option, dir);
+ *
+ * @brief handle the logic related to DHCP option 53
+ *
+ * @param context       Device (interface) context
+ * @param dhcp_option   pointer to DHCP option buffer space
+ * @param dir           packet direction
+ *
+ * @return none
+ */
+static void handle_dhcp_option_53(dhcp_device_context_t *context, const u_char *dhcp_option, dhcp_packet_direction_t dir)
+{
+    switch (dhcp_option[2])
+    {
+    case 1:
+        context->counters[dir].discovery++;
+        if ((context->is_uplink && dir == DHCP_TX) || (!context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].discovery, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 2:
+        context->counters[dir].offer++;
+        if ((!context->is_uplink && dir == DHCP_TX) || (context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].offer, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 3:
+        context->counters[dir].request++;
+        if ((context->is_uplink && dir == DHCP_TX) || (!context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].request, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 5:
+        context->counters[dir].ack++;
+        if ((!context->is_uplink && dir == DHCP_TX) || (context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].ack, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 6 ... 8:
+        // options NAK, Release, Inform
+        break;
+    default:
+        syslog(LOG_WARNING, "pcap_callback(%s): Unknown DHCP option 53 type %d", context->intf,
```

Comment:
Case 4 (DHCPDECLINE) is not handled by this switch, thus it will fall into the default  and get logged as an unknown type, when it should instead be ignored silently like cases 6 through 8.

Reply:
nice catch! done